interfaces/builtin/block_devices.go

```diff
@@ -53,14 +53,15 @@ const blockDevicesConnectedPlugAppArmor = `
 /sys/devices/**/block/** r,
 
 # Access to raw devices, not individual partitions
-/dev/hd[a-t] rw,                         # IDE, MFM, RLL
-/dev/sd{,[a-h]}[a-z] rw,                 # SCSI
-/dev/sdi[a-v] rw,                        # SCSI continued
-/dev/i2o/hd{,[a-c]}[a-z] rw,             # I2O hard disk
-/dev/i2o/hdd[a-x] rw,                    # I2O hard disk continued
-/dev/mmcblk[0-9]{,[0-9],[0-9][0-9]} rw,  # MMC (up to 1000 devices)
-/dev/nvme[0-9]{,[0-9]} rw,               # NVMe (up to 100 devices)
-/dev/vd[a-z] rw,                         # virtio
+/dev/hd[a-t] rw,                                          # IDE, MFM, RLL
+/dev/sd{,[a-h]}[a-z] rw,                                  # SCSI
+/dev/sdi[a-v] rw,                                         # SCSI continued
+/dev/i2o/hd{,[a-c]}[a-z] rw,                              # I2O hard disk
+/dev/i2o/hdd[a-x] rw,                                     # I2O hard disk continued
+/dev/mmcblk[0-9]{,[0-9],[0-9][0-9]} rw,                   # MMC (up to 1000 devices)
+/dev/nvme{[0-9],[1-9][0-9]} rw,                           # NVMe (up to 100 devices)
+/dev/nvme{[0-9],[1-9][0-9]}n{[1-9],[1-5][0-9],6[0-3]} rw, # NVMe (up to 100 devices, with 1-63 namespaces)
```

Comment:
As mentioned in the forum, we need this rule for sure, but also a change to snap-device-helper. From the post:
"
There is a bug in snap-device-helper that adds the device with c 259:0 rwm rather than b 259:0 rwm and the fix is to change this:
```
if [ "${DEVPATH#*/block/}" != "$DEVPATH" ]; then
    type="b"
else
    type="c"
fi
```
to be:
```
if [ "${DEVPATH#*/block/}" != "$DEVPATH" ]; then
    type="b"
elif [ "${DEVPATH#*/nvme/}" != "$DEVPATH" ]; then
    type="b"
else
    type="c"
fi
```
"
While this might be more justification for rewriting snap-device-helper as a small C program (C because an exploratory go binary was quite large, all udev helpers are teeny and we could perhaps share some code with snap-confine), that should *not* been done in this PR. There are different ways to fix snap-device-helper than the above, but it is clear and has the same performance characteristics as what we have now. It is also easy to verify the behavior:
```
$ sudo mkdir /sys/fs/cgroup/devices/snap.test.me
$ sudo sh -c 'echo "a *.* rwm" > /sys/fs/cgroup/devices/snap.test.me/devices.deny'
$ count=0 ; for i in /devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.0/host8/target8:0:0/8:0:0:0/block/sda/sda1 /devices/pci0000:00/0000:00:1d.0/0000:3c:00.0/nvme/nvme0/nvme0n1 /devices/virtual/mem/null ; do count=$((count+1)) ; sudo /usr/lib/snapd/snap-device-helper add snap_test_me $i $count:$count ; done
$ cat /sys/fs/cgroup/devices/snap.test.me/devices.list
b 1:1 rwm
b 2:2 rwm  # this is what we fixed (was 'c 2:2 rwm')
c 3:3 rwm
```
(which could be adapted to a spread test).

Reply:
This patch is not quite right, because we have some devpaths with "nvme" in it that match to block devices and some that match to char devices. I ended up with something slightly more complicated, but I still want to write a spread test like this but have not been able to do so yet unfortunately...

Reply:
> This patch is not quite right, because we have some devpaths with "nvme" in it that match to block devices and some that match to char devices. I ended up with something slightly more complicated, but I still want to write a spread test like this but have not been able to do so yet unfortunately...

Thanks for diving deeper on that. :)

Reply:
I wrote a basic spread test which tests for both types of nvme device